### ISSUE TITLE
You can't shoot on the go

### ISSUE THREAD

**Issue by DmitryHetman** [rainautumn/ttytanks#10]

**Comment by rainautumn** [comment]
А это идея! основной поток сделать зацикленным на то чтобы собирать кнопки нажатые. ты об этом?

**Comment by DmitryHetman** [comment]
Да, ну Я не знаю как реализовывается такое, думал считывать надо нажатие в отдельном потоке, и стрелять, или может evdev или libinput юзать

**Comment by DmitryHetman** [comment]
Но кнопки управления лучше с помошью case считывать, нельзя в две стороны одновременно ехать, отдельный поток нужен только для стрельбы

**Comment by rainautumn** [comment]
ну это мелочь я вот думаю как протокол разработать для p2p игры.

**Comment by DmitryHetman** [comment]
Я думаю уже все должно быть где-нибудь, библиотека какая-то, или торрент клиент.
Если p2p делать то надо установить связьежду клиентами, а после этого простотранслировать команды на машину соперника, это за одно и защита от читеров будет ибо кто победил решается на твоей машине

**Comment by DmitryHetman** [comment]
IRC канал можно завести для обсуждения игры, наверно надо будет завести.
.

**Comment by DmitryHetman** [comment]
Но лучше не торопись, для начала надо наладить саму игру, кстати в ANSI символах нашел символы более подходящие под границу геймспейса, двойные полосы

**Comment by DmitryHetman** [comment]
Код надо бы упростить, некоторые строчки в render.c непонятны, для чего надо math.h не очень ясно, игра простая вроде.

**Comment by rainautumn** [comment]
ну я планировал stun заюзать. и смысл брать настолько готовые модули? можно было просто скачать готовую игру

**Comment by DmitryHetman** [comment]
Я не знаю как с сетями работать, главное чтобы работало, но для начала надо довести игру с ботами до рабочего состояния

**Comment by DmitryHetman** [comment]
Было бы лучше если бы боты не убивали друг друга

**Comment by rainautumn** [comment]
ну у ботов отдельные пули же. можно реализовать вроде

**Comment by DmitryHetman** [comment]
Что-то код у тебя не лаконичный, недавно сишку учить начал? Наверно стоило бы почитать чужой код, терминальный игр много, и вообще всего.

**Comment by DmitryHetman** [comment]
Я вот смотрю много чего повторяется в коде, может быть стоило переписать все в функциональном стиле? Количество фунций станет больше, но количество кода значительно уменьшится должно. Но перед этим бы то что есть понять

**Comment by rainautumn** [comment]
это была первая программа на си
делал ctrl c ctrl v каюсь

**Comment by DmitryHetman** [comment]
Надо делать это функциями отдельными, хоть так будет чуть медленнее работать, оптимизации  -О2 должны все компенсировать инлайном функций в бинарнике.

**Comment by DmitryHetman** [comment]
А сам ввод с клавиатуры нельзя через getch реализовывать, оно зависит от скорости терминала, так не годится, надо считывать событие что нажата такая-то клавиша, и пока она не выключена что-то делать и задержка, так будет одинаково везде работать.

**Comment by rainautumn** [comment]
ну логично.. возможно поищю варианты в свободное время

**Comment by DmitryHetman** [comment]
Без этого о мультиплеере можно забыть.

**Comment by rainautumn** [comment]
ну да. логично. условия будут сильно не равны

**Comment by rainautumn** [comment]
но мне интересна безсерверная реализация мультиплеера в подобной игре.

**Comment by DmitryHetman** [comment]
Для начала прийдется перейти на evdev /libevdev  или libinput для ввода.
Реализация довольно проста, надо только установить соединение и обмениватся нажатыми клавишами, если сейчас мультиплеер прикрутить то будет обмен введенными смэимволами, а их количество в секунду оочень зависимо от терминала, скорость реакции тоже от терминалазависит(задержка перед повторением символов)

**Comment by DmitryHetman** [comment]
Убери баг в  boom_render, тот цикл что в начале, переменная i если ее сократить то количество пуль ограничится, я вообще не понял что это значит и как это работает, кстати используй скрипт make_statistic после игры, из игры надо по нормальномувыйти перед этим, Q- выход из самой игры, 0 выход из меню. запусти скрипт, в папке со статистикой появятся файлы со статистикой используемых функций и строк кода, самое используемое надо оптимизировать.

**Comment by DmitryHetman** [comment]
И не принимай реквесты в main

**Comment by rainautumn** [comment]
черт. хотел в дев

**Comment by rainautumn** [comment]
забыл

**Comment by DmitryHetman** [comment]
Я забыл что у тебя дев и кинул в мастера.

**Comment by DmitryHetman** [comment]
Вообще лучше было бы помимо дев создать отдельные микроветки для готовых бинарей,  соурцес, лайтсоурцес, и файлов в корне

**Comment by rainautumn** [comment]
ну черт. я чет думаю о подобном протоколе как в торрентах. но чет не найду поднятого stun сервака. уже nmap стер. ты не видел stun сервака живого?

**Comment by DmitryHetman** [comment]
надо запихнуть скрипты в папку scripts и чтобы запускались из мейкфайла. Я уже знаю как это делается.

**Comment by rainautumn** [comment]
а по поводу микроветок.. стоит ли е*****ся с мержами из за 1000 строк и полутора контрибьюторов?

**Comment by DmitryHetman** [comment]
Ты говорил мог по Lan сделать мультиплеер, сделай без сервера, по этому же принципу через интернет прикрутишь, можно по локалкесделать?

**Comment by DmitryHetman** [comment]
Может и не стоет но это хороший тон, лучше сразу привыкнуть чем переучиватся.

**Comment by rainautumn** [comment]
не не не. я хочу через nat делать комнаты. В логалка сделать сеть слишком легко..
привыкать? просто когда 4-5 пишут одновременно то ветки по таскам 100% стоит делать
а когда двое можно и не ветвить.

**Comment by DmitryHetman** [comment]
Да и мои пулл-реквесты достаточно громоздки, а так  отдельно каждый файл  отдельный реквест, это удобно с одной стророны, а сдругой, если много изменений то код может не работать пока все не приймешь.

**Comment by rainautumn** [comment]
ну тут проблема что у тебя нет доступа к репозиторию. а ядажене знаю.. гитхаб умеет делать так?

**Comment by rainautumn** [comment]
и вообще нужно лиэто

**Comment by DmitryHetman** [comment]
Я без понятия о этих либах и протоколах, начать надо с чего-то. Начни с исправления ввода и boom_render, был бы рад если бы ты объяснил как она работает и как используется.
После этих фиксов делай мультиплеер по локалке,  его в lan_multilplayer.c пиши если он будет сильно от сетевого отличатся.

**Comment by DmitryHetman** [comment]
Он все умеет, но в нашем случае может и не нужно, если большой коммит то приняв отдельный реквест ты можешь сломать все.

**Comment by DmitryHetman** [comment]
роль render all тоже не очень ясна

**Comment by rainautumn** [comment]
//i - is not a distance for bullet to fly, it's number of bullets that you can shoot

**Comment by rainautumn** [comment]
ты все верно откоментил

**Comment by DmitryHetman** [comment]
Это так, но Я не понял до конца, когда оно равно 4  то будет 4 выпущеных пули и все, если столько сколько сейчас то количество бесконечно, хотя там константа.
Строчки с abs() непонятны,.
Эта функция за раз всем танкам рисует патрон?

**Comment by rainautumn** [comment]
да всем танкам рисует бесконечно патронов

**Comment by DmitryHetman** [comment]
И сколько раз этот цикл повторяется и почему

**Comment by rainautumn** [comment]
1 раз за прорисовку выполняется

**Comment by DmitryHetman** [comment]
Я кажись понял как улучшить это, вместо i++  вписать i+=2, таким образом будет шаг в одну клетку? Надо же зажержка между выстрелами.

**Comment by rainautumn** [comment]
нет. тогда каждый второй выстрел не выстрелит. не так радикально то

**Comment by rainautumn** [comment]
тут только отфиксить ввод с клавиатуры

**Comment by DmitryHetman** [comment]
Эффекивнее было бы как-то иначе все это сделать, при каждом выстреле создавать поток в котором запускается функция стрельбы. Может лучше сделать отдельные функции для стрельбы вверх вниз влево вправо

**Comment by rainautumn** [comment]
кстати а в С++11 99 17 нет ничего по вводу с клавы?

**Comment by rainautumn** [comment]
а профит от ухода от абстрагирования? даже не такого годного но абстрагирования.

**Comment by DmitryHetman** [comment]
Ввод через evdev в ядре риализован, libevdev.h либа для обработки сионалов. Вроде то что надо, но хелловорды на ней не очень ясны мне. Игра кстати уже на gnu11 написана, так что можно гнушные дополнения использовать.
Если хочешь откатится на анси11 то напиши вмейкфайле c11 и пофикси ворнинг, там одна фунция гнушная юзалась, пришлось на гну перейти чтобы без ошибоксобирать.

**Comment by DmitryHetman** [comment]
Твоя функция - слишком комбайн, по сути это 4 или даже 8 функций впихнутых в одну.
Вызываешь ее для одной цели, а оно целую кучу проверяет.

**Comment by DmitryHetman** [comment]
А еще она нечитабельна,короткие простые функции- вот что надо.

**Comment by DmitryHetman** [comment]
Я когда ее вызываю просто незнаю что ожидать, она слишком сложная, надо придерживатся unix-way все-таки, новичек обосрется читать этот код.

**Comment by rainautumn** [comment]
ну да. логично.
мне стоит почитать книженок о красивом коде...
ну я ща разгребу сессию попишу.
по сути 3 таска висит 
1-- фикс клавы
2 -- сетка
3  -- звук и фичи до состояния годноты и релизика. 
просто я ща кпч горю.
так вопрос про асечку. дай мне что-нибудь свое.

**Comment by DmitryHetman** [comment]
Хорошо, давай лучше поднимешь IRC chanel в freenode? Так будет  доступная желающим помочь конференция, канал надо в about  добавить.

**Comment by DmitryHetman** [comment]
Я в качестве лички хотел xmpp юзать, но оно ксерверу не конектится, а в чем дело неэнаю

**Comment by rainautumn** [comment]
сентябрь на дворе. 
icq отрефакторили но она та же icq!
хоть я и потерял 6тизначник но она того стоит

**Comment by DmitryHetman** [comment]
Клиенты жирноваты вроде, она централизирована

**Comment by DmitryHetman** [comment]
Я могу с телефона канал открыть, но он у меня вырубается (телефон)

**Comment by DmitryHetman** [comment]
Надо канал для разработчиков открыть

**Comment by rainautumn** [comment]
да надо решить с общением. 
может slack поднять?

**Comment by DmitryHetman** [comment]
Что такое слак?

**Comment by DmitryHetman** [comment]
IRC стандартное и простое  решение для этой задачи

**Comment by DmitryHetman** [comment]
В  интернете должны быть хвлявные сервера или может это встроено в freenode чтобы канал не закрывался как у тебя клиент отключится

**Comment by DmitryHetman** [comment]
Если личку то xmpp, Я толком только не понял как один клиент использовать.

**Comment by rainautumn** [comment]
icq купила mail.ru расходимся.
slack.com же

**Comment by rainautumn** [comment]
давай IRC

**Comment by DmitryHetman** [comment]
IRC Очень легко поднять

**Comment by DmitryHetman** [comment]
Если у тебя стабильный интернет то потом сам поднимешь сервер

**Comment by rainautumn** [comment]
https://webchat.freenode.net/ я чет не вникаю.. ttytanks вводишь тут и ты в irc?

**Comment by DmitryHetman** [comment]
freenode
имя канала ttytanks

**Comment by DmitryHetman** [comment]
Интернет у меня пока дерьмовый, есть Bip который вроде безлимитный

**Comment by DmitryHetman** [comment]
Можешь через Bip если че связаться, нормер +380939417234

**Comment by DmitryHetman** [comment]
If launch tty from root you can read events directly from linux kernel. We should use some library on top of it.
